tests/protocols/test_websocket.py

```diff
@@ -597,9 +597,8 @@ async def connect(url):
         port=unused_tcp_port,
     )
     async with run_server(config):
-        with pytest.raises(websockets.exceptions.ConnectionClosed) as exc_info:
+        with pytest.raises(websockets.exceptions.WebSocketException):
             await connect(f"ws://127.0.0.1:{unused_tcp_port}")
-        assert exc_info.value.code == 1006
```

Comment:
Is the behavior different between wsproto and websockets implementation different here?

Reply:
I believe that they do. I will double check how the 4 parametrized variants of this test behave and report back.

Reply:
I'd like to avoid having a different behavior, but it's not a blocker. If we do, it would be cool that the test would show how they differ (it can be without the parametrize and create a new test).

Reply:
Understood. I'll see what I can do.

Reply:
We get the following errors in the test client:

- wsproto: `ConnectionClosedError(None, None, None)`
- websockets 10.4: `ConnectionClosedError(None, None, None)`
- websockets 11:
    - before this PR: `ConnectionClosedOK(Close(code=1000, reason=''), Close(code=1000, reason=''), True)`
    - after this PR: `InvalidMessage('did not receive a valid HTTP response')`

This is happening because:

- `WebSocketProtocol.run_asgi` sets `self.closed_event` here: https://github.com/encode/uvicorn/blob/8239373ec649b1a41179070b6ba2c8dc549df63e/uvicorn/protocols/websockets/websockets_impl.py#L258
- `WebSocketProtocol.ws_handler` exits here: https://github.com/encode/uvicorn/blob/8239373ec649b1a41179070b6ba2c8dc549df63e/uvicorn/protocols/websockets/websockets_impl.py#L239
- As soon as the connection handler returns, websockets performs the closing handshake automatically.

The last step used to be delayed by one cycle of the event loop in websockets 10.4. This used to allow `WebSocketProtocol.run_asgi` to close the transport first here: https://github.com/encode/uvicorn/blob/8239373ec649b1a41179070b6ba2c8dc549df63e/uvicorn/protocols/websockets/websockets_impl.py#L268

Reply:
I am updating the PR with a small change that restores the previous behavior.